torch/fx/graph.py

```diff
@@ -151,36 +163,38 @@ def create_name(self, candidate: str, obj: Optional[Any]) -> str:
         if obj is not None and obj in self._obj_to_name:
             return self._obj_to_name[obj]
 
-        # delete all characters that are illegal in a Python identifier
-        candidate = self._illegal_char_regex.sub("_", candidate)
+        # optimistically check if candidate is already a valid name
+        match = _name_regex.match(candidate)
+        if match is None:
+            # delete all characters that are illegal in a Python identifier
+            candidate = _illegal_char_regex.sub("_", candidate)
 
-        if not candidate:
-            candidate = "_unnamed"
+            if not candidate:
+                candidate = "_unnamed"
 
-        if candidate[0].isdigit():
-            candidate = f"_{candidate}"
+            if candidate[0].isdigit():
+                candidate = f"_{candidate}"
 
-        match = self._name_suffix_regex.match(candidate)
-        if match is None:
-            base = candidate
-            num = None
-        else:
-            base, num_str = match.group(1, 2)
-            num = int(num_str)
+            match = _name_regex.match(candidate)
+            assert match is not None
 
-        candidate = base if num is None else f"{base}_{num}"
-        if not num:
-            num = self._base_count[base]
+        base, num = match.group(1, 2)
+        if num is None or candidate in self._used_names:
+            num = self._base_count.get(candidate, 0)
+            if _illegal_names.get(candidate, obj) is not obj:
+                num += 1
+                candidate = f"{base}_{num}"
+                # assume illegal names don't end in _\d so no need to check again
+        else:
+            num = int(num)
 
-        while candidate in self._used_names or self._is_illegal_name(candidate, obj):
+        while candidate in self._used_names:
             num += 1
             candidate = f"{base}_{num}"
 
         self._used_names.add(candidate)
         self._base_count[base] = num
-        if obj is None:
-            self._unassociated_names.add(candidate)
-        else:
+        if obj is not None:
             self._obj_to_name[obj] = candidate
         return candidate
```

Comment:
May want to want to add an `assert candidate.isidentifier()` for good measure. :)

Reply:
This PR is about performance optimization, that would make things slower and would be redundant with the regexp.